File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/rest-api/src/main/java/RestApi.java
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class RestApi {
    private final Map<String, User> users;

    RestApi(User... users) {
        this.users = new HashMap<>();
        for (User user : users) {
            this.users.put(user.name(), user);
        }
    }

    String get(String url) {
        if ("/users".equals(url)) {
            return getAllUsers();
        }
        return "";
    }

    String get(String url, JSONObject payload) {
        if ("/users".equals(url)) {
            return getSpecificUsers(payload);
        }
        return "";
    }

    String post(String url, JSONObject payload) {
        if ("/add".equals(url)) {
            return addUser(payload);
        } else if ("/iou".equals(url)) {
            return addIou(payload);
        }
        return "";
    }

    private String getAllUsers() {
        List<User> userList = new ArrayList<>(users.values());
        userList.sort(Comparator.comparing(User::name));
        return toUsersJson(userList);
    }

    private String getSpecificUsers(JSONObject payload) {
        List<String> requestedNames = payload.getJSONArray("users").toList()
            .stream().map(Object::toString).collect(Collectors.toList());
        
        List<User> userList = users.values().stream()
            .filter(user -> requestedNames.contains(user.name()))
            .sorted(Comparator.comparing(User::name))
            .collect(Collectors.toList());
            
        return toUsersJson(userList);
    }

    private String addUser(JSONObject payload) {
        String name = payload.getString("user");
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("User name cannot be null or empty");
        }
        if (users.containsKey(name)) {
            throw new IllegalArgumentException("User already exists");
        }
        User newUser = User.builder().setName(name).build();
        users.put(name, newUser);
        return newUser.toJson().toString();
    }

    private String addIou(JSONObject payload) {
        String lender = payload.getString("lender");
        String borrower = payload.getString("borrower");
        double amount = payload.getDouble("amount");

        if (lender == null || borrower == null) {
            throw new IllegalArgumentException("Lender and borrower cannot be null");
        }
        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be positive");
        }
        if (!users.containsKey(lender)) {
            throw new IllegalArgumentException("Lender does not exist");
        }
        if (!users.containsKey(borrower)) {
            throw new IllegalArgumentException("Borrower does not exist");
        }

        User lenderUser = users.get(lender);
        User borrowerUser = users.get(borrower);

        // Update lender's owedBy
        User updatedLender = User.builder()
            .setName(lender)
            .owedBy(borrower, amount)
            .build();

        // Update borrower's owes
        User updatedBorrower = User.builder()
            .setName(borrower)
            .owes(lender, amount)
            .build();

        // Merge with existing data
        users.put(lender, mergeUsers(lenderUser, updatedLender));
        users.put(borrower, mergeUsers(borrowerUser, updatedBorrower));

        List<User> result = List.of(users.get(lender), users.get(borrower));
        result.sort(Comparator.comparing(User::name));
        return toUsersJson(result);
    }

    private User mergeUsers(User existing, User updates) {
        User.Builder builder = User.builder().setName(existing.name());
        
        // Add existing owes
        for (Iou iou : existing.owes()) {
            builder.owes(iou.name(), iou.amount());
        }
        // Add update owes
        for (Iou iou : updates.owes()) {
            builder.owes(iou.name(), iou.amount());
        }
        
        // Add existing owedBy
        for (Iou iou : existing.owedBy()) {
            builder.owedBy(iou.name(), iou.amount());
        }
        // Add update owedBy
        for (Iou iou : updates.owedBy()) {
            builder.owedBy(iou.name(), iou.amount());
        }
        
        return builder.build();
    }

    private String toUsersJson(List<User> users) {
        JSONObject result = new JSONObject();
        result.put("users", users.stream()
            .map(User::toJson)
            .collect(Collectors.toList()));
        return result.toString();
    }

}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/rest-api/rest-api.js
export class RestAPI {
  constructor(database = { users: [] }) {
    this.users = database.users.reduce((acc, user) => {
      acc[user.name] = {
        ...user,
        owes: user.owes || {},
        owed_by: user.owed_by || {},
        balance: (user.owed_by ? Object.values(user.owed_by).reduce((a, b) => a + b, 0) : 0) -
                 (user.owes ? Object.values(user.owes).reduce((a, b) => a + b, 0) : 0)
      };
      return acc;
    }, {});
  }

  get(url) {
    const [path, params] = url.split('?');
    if (path === '/users') {
      const query = params ? new URLSearchParams(params) : null;
      const names = query ? query.get('users')?.split(',') : null;
      const users = names ? 
        names.map(name => this.getUser(name)).filter(Boolean) :
        Object.values(this.users);
      return {
        users: users.sort((a, b) => a.name.localeCompare(b.name))
      };
    }
    throw new Error('Invalid endpoint');
  }

  post(url, payload) {
    switch (url) {
      case '/add':
        return this.addUser(payload.user);
      case '/iou':
        return this.addIOU(payload.lender, payload.borrower, payload.amount);
      default:
        throw new Error('Invalid endpoint');
    }
  }

  getUser(name) {
    const user = this.users[name];
    if (!user) return null;
    return {
      name: user.name,
      owes: { ...user.owes },
      owed_by: { ...user.owed_by },
      balance: user.balance
    };
  }

  addUser(name) {
    if (this.users[name]) {
      throw new Error('User already exists');
    }
    this.users[name] = {
      name,
      owes: {},
      owed_by: {},
      balance: 0
    };
    return this.getUser(name);
  }

  addIOU(lender, borrower, amount) {
    if (!this.users[lender] || !this.users[borrower]) {
      throw new Error('User not found');
    }

    // Update lender's owed_by
    this.users[lender].owed_by[borrower] = 
      (this.users[lender].owed_by[borrower] || 0) + amount;
    this.users[lender].balance += amount;

    // Update borrower's owes
    this.users[borrower].owes[lender] = 
      (this.users[borrower].owes[lender] || 0) + amount;
    this.users[borrower].balance -= amount;

    // Simplify debts if possible
    const lenderOwed = this.users[lender].owed_by[borrower] || 0;
    const borrowerOwes = this.users[borrower].owes[lender] || 0;
    
    if (lenderOwed > 0 && borrowerOwes > 0) {
      if (lenderOwed > borrowerOwes) {
        this.users[lender].owed_by[borrower] -= borrowerOwes;
        delete this.users[borrower].owes[lender];
      } else if (lenderOwed < borrowerOwes) {
        this.users[borrower].owes[lender] -= lenderOwed;
        delete this.users[lender].owed_by[borrower];
      } else {
        delete this.users[lender].owed_by[borrower];
        delete this.users[borrower].owes[lender];
      }
    }

    return {
      users: [this.getUser(lender), this.getUser(borrower)]
        .sort((a, b) => a.name.localeCompare(b.name))
    };
  }
}
